third_party/move/tools/move-unit-test/src/lib.rs

```diff
@@ -218,8 +228,31 @@ impl UnitTestingConfig {
         genesis_state: Option<ChangeSet>,
         cost_table: Option<CostTable>,
         writer: W,
+    ) -> Result<(W, bool)> {
+        self.run_and_report_unit_tests_with_gas_meter(
+            test_plan,
+            native_function_table,
+            genesis_state,
+            writer,
+            GasStatus::new(
+                &cost_table.unwrap_or_else(unit_cost_table),
+                Gas::new(self.gas_limit.unwrap_or(DEFAULT_EXECUTION_BOUND)),
+            ),
+        )
+    }
+
+    /// Public entry point to Move unit testing as a library
+    /// Returns `true` if all unit tests passed. Otherwise, returns `false`.
+    pub fn run_and_report_unit_tests_with_gas_meter<W: Write + Send, G: TestGasMeter + Send>(
+        &self,
+        test_plan: TestPlan,
+        native_function_table: Option<NativeFunctionTable>,
+        genesis_state: Option<ChangeSet>,
+        writer: W,
+        gas_meter: G,
     ) -> Result<(W, bool)> {
         let shared_writer = Mutex::new(writer);
+        let shared_gas_meter = Mutex::new(gas_meter);
```

Comment:
It does not make sense to have a shared gas meter. Each test should have their own. 

I think what we need here is some sort of gas meter factory.